kratos/utilities/geometrical_sensitivity_utility.cpp

```diff
@@ -0,0 +1,282 @@
+
+    void CalculateSensitivity(IndexType iNode, IndexType iCoord, double& rDetJ_Deriv, ShapeFunctionsGradientType& rDN_DX_Deriv) const;
+
+    ///@}
+
+private:
+    ///@name Member Variables
+    ///@{
+
+    const JacobianType& mrJ;
+    const ShapeFunctionsLocalGradientType& mrDN_De;
+    MatrixType mCofactorJ;
+    double mDetJ;
+
+    ///@}
+    ///@name Private Operations
+    ///@{
+
+    template<class TMatrixType>
+    double CalculateDeterminant(const TMatrixType& rMat) const;
+
+    template<class TMatrixType>
+    double CalculateCofactor(const TMatrixType& rMat, IndexType i, IndexType j) const;
+
+    template<class TMatrixType>
+    MatrixType CalculateCofactorMatrix(const TMatrixType& rMat) const;
+
+    double CalculateDeterminantOfJacobianSensitivity(IndexType iNode, IndexType iCoord) const;
+
+    MatrixType CalculateCofactorOfJacobianSensitivity(IndexType iNode, IndexType iCoord) const;
+
+    ///@}
+};
+
+///@} // Kratos Classes
+///@} // Adjoint Fluid Application group
+
+GeometricalSensitivityUtility::GeometricalSensitivityUtility(const JacobianType& rJ, const ShapeFunctionsLocalGradientType& rDN_De)
+: mpImpl(new Impl(rJ, rDN_De))
+{
+    KRATOS_TRY;
+    mpImpl->Initialize();
+    KRATOS_CATCH("");
+}
+
+GeometricalSensitivityUtility::~GeometricalSensitivityUtility() = default;
+
+void GeometricalSensitivityUtility::CalculateSensitivity(IndexType iNode, IndexType iCoord, double& rDetJ_Deriv, ShapeFunctionsGradientType& rDN_DX_Deriv) const
+{
+    KRATOS_TRY;
+    mpImpl->CalculateSensitivity(iNode, iCoord, rDetJ_Deriv, rDN_DX_Deriv);
+    KRATOS_CATCH("");
+}
+
+void GeometricalSensitivityUtility::Impl::Initialize()
+{
+    KRATOS_TRY;
+
+    KRATOS_ERROR_IF(mrJ.size1() != mrJ.size2()) << "Non-square Jacobian matrix." << std::endl;
+
+    KRATOS_ERROR_IF(mrJ.size2() != mrDN_De.size2())
+        << "Jacobian local-coordinates size (" << mrJ.size2()
+        << ") != shape function local-coordinates size(" << mrDN_De.size2()
+        << ")." << std::endl;
+
+    mCofactorJ = CalculateCofactorMatrix(mrJ);
+    mDetJ = CalculateDeterminant(mrJ);
+
+    KRATOS_CATCH("");
+}
+
+void GeometricalSensitivityUtility::Impl::CalculateSensitivity(IndexType iNode, IndexType iCoord, double& rDetJ_Deriv, ShapeFunctionsGradientType& rDN_DX_Deriv) const
+{
+    KRATOS_TRY;
+
+    rDetJ_Deriv = CalculateDeterminantOfJacobianSensitivity(iNode, iCoord);
+
+    MatrixType cofactorJ_deriv = CalculateCofactorOfJacobianSensitivity(iNode, iCoord);
+    if (rDN_DX_Deriv.size1() != mrDN_De.size1() || rDN_DX_Deriv.size2() != mCofactorJ.size1())
+        rDN_DX_Deriv.resize(mrDN_De.size1(), mCofactorJ.size1());
+    noalias(rDN_DX_Deriv) = (1.0 / mDetJ) * prod(mrDN_De, trans(cofactorJ_deriv));
+    rDN_DX_Deriv += -(rDetJ_Deriv / (mDetJ * mDetJ)) * prod(mrDN_De, trans(mCofactorJ));
+
+    KRATOS_CATCH("");
+}
+
+template<class TMatrixType>
+double GeometricalSensitivityUtility::Impl::CalculateDeterminant(const TMatrixType& rMat) const
```

Comment:
There is a determinant calculation function in math_utils that is very similar to this, except that it only supports bounded matrices. I'd suggest replacing it by this implementation, since this one is more flexible.

Reply:
I would prefer not to change the function in math_utils because it has a safer interface than the one I am using. I would need to implement several runtime checks on the dimensions and datatype for the more general version. In my case, it's in the private section so I have control over who uses it and how it is called.

Reply:
maybe I can do it with minimal checking. Boost already checks for bad index and something like a decltype might work as a static assert on othe data type of the matrix.